pkg/venafi/tpp/connector.go

```diff
@@ -1231,31 +1265,53 @@ func (c *Connector) getIdentity(userName string) (*policy.IdentityEntry, error)
 		return nil, err
 	}
 
-	return c.getIdentityMatching(resp.Identities, userName)
-}
+	// When TPP looks for a username that matches the filter, an implicit
+	// wildcard is added to the end of the filter string. For example, imagining
+	// that `jsmith` and `jsmithson` are existing identities, searching for
+	// `jsmith` will return both `jsmith` and `jsmithson`. In the case of local
+	// identities, `jsmith` will always be returned first. But in the case of AD
+	// and LDAP, the order of these results may be different, and `jsmithson`
+	// may be unexpectedly returned first. This same problem may appear when an
+	// AD or LDAP provider has been configured to access the local identities:
+	// `jsmithson` may get returned first if `jsmithson` only exists in AD
+	// (because the AD results are returned before the local identities).
+	//
+	// The wildcard problem only affects usernames, not emails. That's because
+	// the LDAP query recommended for enabling user search by email in the
+	// Venafi Configuration Console is based on exact match, unlike `anr` used
+	// for searching usernames. Thus, we do not need to check for an exact match
+	// when an email is provided.
 
-func (c *Connector) getIdentityMatching(identities []policy.IdentityEntry, identityName string) (*policy.IdentityEntry, error) {
-	var identityEntryMatching *policy.IdentityEntry
+	// When the filter isn't an email, we consider that it is a username.
+	_, err = mail.ParseAddress(filter)
+	isEmail := err == nil
+	isUsername := !isEmail
 
-	if len(identities) > 0 {
-		for i := range identities {
-			identityEntry := identities[i]
-			if identityEntry.Name == identityName {
-				identityEntryMatching = &identityEntry
-				break
+	switch {
+	case len(resp.Identities) == 0:
+		return nil, fmt.Errorf("no identity found for '%s'", filter)
+	case len(resp.Identities) == 1:
+		return &resp.Identities[0], nil
+	case len(resp.Identities) > 1 && isUsername:
+		// The username case: we need to ignore the results that are prefixes of
+		// the queried username. For example, if the filter is `jsmith`, we
+		// ignore `jsmithson` and `jsmithers`.
+		for _, identity := range resp.Identities {
+			if identity.Name == filter {
+				return &identity, nil
 			}
 		}
-	}
-
-	//if the identity is not null
-	if identityEntryMatching != nil {
-		return identityEntryMatching, nil
-	} else {
-		return nil, fmt.Errorf("it was not possible to find the user %s", identityName)
+		return nil, fmt.Errorf("unexpected: browseIdentities(%q) returned %d identities but none of them match the username exactly", filter, len(resp.Identities))
```

Comment:
I'm not sure if we should update the error message because we don't know if there are VCert users expecting the current error message, so for them this can cause an compatibility issue.

Reply:
I'm not opposed to change the error message in favor of better understanding of the error for better UX.
I'd change this part:
`...identities but none of them match the username exactly`
to
`...identities but none of them match the username %s`
where %s is the username that was trying to be filtered. 

I'd say that filtering by username, should be assumed that value have to be exact.

Another way to highlight that value should be exact would be adding quotes:
`...identities but none of them match the username "%s"`

Reply:
I liked the new proposal for the error message :+1: